src/ckb/channel.rs

```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+            self.get_id(),
+            &detailed_tlc,
+            self.to_local_amount,
+            self.to_remote_amount
+        );
+        Ok(detailed_tlc)
+    }
+
+    // Remove a tlc with a reason. If the tlc is removed, then the channel
+    // balance will be updated accordingly. Otherwise, it is guaranteed that
+    // the channel state is not updated.
+    pub fn remove_tlc_with_reason(
+        &mut self,
+        tlc_id: TLCId,
+        reason: RemoveTlcReason,
+    ) -> Result<DetailedTLCInfo, ProcessingChannelError> {
+        let commitment_numbers = self.get_current_commitment_numbers();
+
+        match self.tlcs.entry(tlc_id) {
+            Entry::Occupied(mut entry) => {
+                let current = entry.get_mut();
+                match current.removal_info {
+                    Some((removed_at, remove_reason))
+                        if removed_at == commitment_numbers && reason == remove_reason =>
+                    {
+                        debug!(
+                            "Skipping removing of tlc {:?} as it is already removed at {:?} with the same reason {:?}"
+                            , current, removed_at, reason
+                        );
+                        return Ok(current.clone());
+                    }
+                    Some((removed_at, remove_reason)) => {
+                        error!("Trying to remove an already removed tlc {:?} (removed at {:?}) with different reason: {:?} (old reason {:?})", current, removed_at, reason, remove_reason);
+                    }
+                    None => {
+                        if let RemoveTlcReason::RemoveTlcFulfill(fulfill) = reason {
+                            let filled_payment_hash: Hash256 =
+                                blake2b_256(fulfill.payment_preimage).into();
+                            if current.tlc.payment_hash != filled_payment_hash {
+                                return Err(ProcessingChannelError::InvalidParameter(format!(
+                                    "Preimage {:?} is hashed to {}, which does not match payment hash {:?}",
+                                    fulfill.payment_preimage, filled_payment_hash, current.tlc.payment_hash,
+                                )));
+                            }
+                        }
+                    }
+                }
+                // We are sure that this tlc removal is not repeated and the reason is valid now.
+                let should_increase_our_balance = match reason {
+                    RemoveTlcReason::RemoveTlcFulfill(_) => tlc_id.is_received(),
+                    RemoveTlcReason::RemoveTlcFail(_) => tlc_id.is_offered(),
+                };
+                if should_increase_our_balance {
+                    self.to_local_amount += current.tlc.amount;
+                } else {
+                    self.to_remote_amount += current.tlc.amount;
+                }
+                current.removal_info = Some((commitment_numbers, reason));
```

Comment:
is it better we remove the current entry from `self.tlcs`?
otherwise `tlcs` will keep growing, and we need to serialize and store this field.

Reply:
We need this historical tlc information to construct valid revocation transaction. Imagine that the adversary broadcasts an older commitment transaction, and we want to revoke this. In order to do so, we will need to fill in all the tlcs that are committed commitment transaction in the witness. This is the reason why we need to keep tlcs forever. Of course, we can also save the witnesses for commitment transaction (just hook into this event https://github.com/nervosnetwork/cfn-node/blob/c1c4557a15d6457b23008328e1fc5cdf82e9611b/src/ckb/network.rs#L152C5-L152C27). I personally prefer the latter method. But I haven't think carefully about the tradeoffs. I think it is not such so important to optimize this yet.